waiter/src/waiter/metrics.clj

```diff
@@ -442,4 +458,73 @@
    :stream-onto-resp-chan (service-timer service-id "stream-onto-resp-chan")
    :stream-read-body (service-timer service-id "stream-read-body")
    :stream-request-rate (service-meter service-id "stream-request-rate")
-   :throughput-meter (service-meter service-id "stream-throughput")})
\ No newline at end of file
+   :throughput-meter (service-meter service-id "stream-throughput")})
+
+(defn update-last-request-time
+  "Updates the last-request-time epoch time in the service-id->last-request-time to the maximum
+   of the current last request time and the provided last-request-datetime in epoch millis."
+  [agent-state service-id ^DateTime candidate-last-request-time]
+  (let [current-last-request-time (get-in agent-state [:pending :service-id->last-request-time service-id])]
+    (if (or (nil? current-last-request-time)
+            (< (.getMillis current-last-request-time)
+               (.getMillis candidate-last-request-time)))
+      (update-in agent-state [:pending :service-id->last-request-time]
+                 assoc service-id candidate-last-request-time)
+      agent-state)))
+
+(defn cleanup-last-request-times
+  "Removes stale entries in service-id->last-request-time based on entries provided in
+   published-service-id->last-request-time. Services who have more recent last request
+   times than those have been published will not be gc-ed."
+  [agent-state correlation-id publish-time
+   published-service-id->last-request-time]
+  (cid/cdebug correlation-id "triggering gc of services")
+  (let [service-id->last-request-time (get-in agent-state [:pending :service-id->last-request-time])
+        service-id->last-request-time'
+        (reduce
+          (fn [acc-service-id->last-request-time [service-id candidate-last-request-time]]
+            (let [current-last-request-time (get acc-service-id->last-request-time service-id)]
+              (cond-> acc-service-id->last-request-time
+                      (and current-last-request-time
+                           (<= (.getMillis current-last-request-time)
+                               (.getMillis candidate-last-request-time)))
+                      (dissoc service-id))))
+          service-id->last-request-time
+          published-service-id->last-request-time)
+        num-items-cleaned (- (count service-id->last-request-time)
+                             (count service-id->last-request-time'))]
+    (when (pos? num-items-cleaned)
+      (cid/cinfo correlation-id "gc-ed" num-items-cleaned "services"))
```

Comment:
Does this get printed every time any single service had a request but doesn't have a subsequent request within the publish time?